### PR TITLE
Improvement/implement repsonsive modal

### LINKED ISSUE #482: Modal is pre-rendered and should be encapsulated from a Dialog
Shouldn't modal html be rendered by JS on the client only on demand?
I would also always place modal markup at the bottom of `<body>`, not anywhere in the whole page - to avoid z-index and html-flow issues.

And so far we have a modal dialog implemented.
I would separate those two components, so that the `dialog` markup is wrapped inside an `modal` block.
And inside a `modal` any component could be viewed in a modal manner.

As I can see at the moment we don't manage the occurrence of multiple modals properly, or parent/child modals, etc.

Compatibility with React, Angular, etc is a concern too.

### PR DESCRIPTION
**Pull request by AndyOGo**
fixes #482 

## Feature Request

* [x] One Modal at a time
* [x] Content source
 * [x] Existing markup
 * [x] Load from external page link, filter by optional CSS-Selector
 * [x] passed ass argument
* [x] Disable body's scrollbars (optional, default enabled)
* [x] Content should be hidden by default and only visible if the modal is active - has `.is-open` class
* [x] Write documentation

## Nice to have
* [ ] React Plug Support (nice to have)

## Modal

* [x] Markup is always appended at the end
* [x] Fullscreen Backdrop
* [x] Encapsulated from content like dialogs, pages, etc.
* [x] Close by:
 * [x] Click outside (backdrop click)
 * [x] Close Button click
 * [x] `Esc`-key pressed
* [x] Modes for display/styles
 * [x] scroll - overflowing content is scrolled (default)
 * [x] fullscreen - modal does not scroll, instead the content takes care of overflow
* [x] Handle Focus states
 * [x] Cache/restore previous focused element
 * [x] avoid tabbing / focusing elements outside of modal block

**Mobile**

**Desktop**

## Dialog

The Dialog puts itself inside a modal container.

**Mobile**

* [x] Fullscreen (no backdrop visible/needed)
* [x] Header can grow if text is too long (multi line wrapping)
* [x] Overflowing content shall scroll
* [x] Footer always at the bottom
 * [x] can grow if content to long

**Desktop**
* [x] Padded Positioning - default `centered`
 * [x] Top
 * [x] Pottom
* [x] Dialog grows as content grows
* [x] Overflowing content shall scroll

### CODE REVIEW

**Comment by sventschui** [comment]
Will review today. Very interested in this one

**Comment by AndyOGo** [comment]
@sventschui 
I'm done with the feature set.

I still need to look over the code and improve it, especially styling is not so clear, because I have no fixed designs. But I think it is pretty good so far.

And the dialog footer needs maybe some more work from what I saw

**Comment by AndyOGo** [comment]
@sventschui 
I just read your feedback again. And I realised that your sheets have responsive position -> desktop = center; mobile = bottom.
I have such a feature include in the dialog (top, middle, bottom). But now I think maybe it would be cool if it is possible to handle this positioning stuff directly in the modal2 component.

**Comment by AndyOGo** [comment]
ready to merge:) just finish docs

**Comment by davidknezic** [comment]
Nice! :) Will have a look.

**Comment by AndyOGo** [comment]
docs added - ready to rumble 😄

File: docs/page/components/dialog.jade [hunk 1]
```diff
@@ -0,0 +1,95 @@
+---
+title: Dialog
+order: 7
+template: page.jade
+tags: terms, conditions, agb, avb
+---
+
+include ./includes/mixins.jade
+
+p.paragraph. A dialog window can be used to present very specific information (such as terms & conditions) to the user.
```
Comment (davidknezic) [inline docs/page/components/dialog.jade]: After a dot, you should continue your text on the next line, with one level of indentation.

File: js/jquery/dialog.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+ * @property {Object} classes - A hash of class names which are rendered by this plugin.
+ * @property {string} [classes.dialog=dialog] - The dialog block class.
+ * @property {string} [classes.header=dialog__header] - The dialogs's backdrop element class.
+ * @property {string} [classes.heading=dialog__heading] - The dialogs's content element class.
+ * @property {string} [classes.close=dialog__close] - The dialogs's close element class.
+ * @property {string} [classes.icon=dialog__close__icon] - The dialogs's close icon element class.
+ * @property {string} [classes.content=dialog__content] - The dialogs's backdrop element class.
+ * @property {string} [classes.footer=dialog__footer] - The dialogs's content element class.
+ * @property {string} [closeAriaLabel=close] - ARIA label for close button.
+ * @property {string} [position=middle] - A string specifying the dialogs vertical position - `'top'`, `'middle'` or `'bottom'`.
```
Comment (davidknezic) [inline js/jquery/dialog.js]: > [position=middle]

that's cool! 😃

File: js/jquery/icon.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+const icon = (path, classes) => `<svg class="icon ${classes}">
+        <use xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="${path}"></use>
+      </svg>`
+
+export default icon
```
Comment (davidknezic) [inline js/jquery/icon.js]: Nice, this is gonna be very helpful!

File: js/jquery/modal2.js [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+ */
+
+/**
+ * This callback triggers before the modal closes.
+ * @callback Modal2~onBeforeClose
+ * @param {Modal2} - The current Modal2 instance.
+ */
+
+/**
+ * This callback triggers after the modal has benn closed.
```
Comment (davidknezic) [inline js/jquery/modal2.js]: > has benn closed

File: js/jquery/modal2.js [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+    if (!this.isOpen) return
+    this.isOpen = false
+
+    this.options.onBeforeClose(this)
+
+    this.unbind()
+
+    this.$modal.remove()
+
+    this.$body.attr('aria-hidden', false)
```
Comment (davidknezic) [inline js/jquery/modal2.js]: Nice, `aria-` attributes are back! 😃

File: docs/page/components/modal2.jade [hunk 6]
```diff
@@ -0,0 +1,153 @@
+---
+title: Modal 2
+order: 7
+template: page.jade
+tags: terms, conditions, agb, avb
+---
+
+include ./includes/mixins.jade
+
+p.paragraph. A modal window can be used to present very specific information (such as terms & conditions) to the user.
```
Comment (davidknezic) [inline docs/page/components/modal2.jade]: Same here

**Comment by davidknezic** [comment]
This pr is impressive! Once again, awesome @AndyOGo! 👏 👏 👏 

I tested and went through the changes. If you agree I'd merge.

Only thing is the weird looking modules part in the auto generated jsdocs. However, I'd insist on fixing that in a separate pr at a later time.

screen shot 2016-09-14 at 17 18 00

**Comment by AndyOGo** [comment]
😄 Thank you 😄 

Yeah I know somehow jsdoc does not realize if a modula and a class are atually the same.
It is caused by the `@exports Modal2`.
I will remove it quickly, just a sec

**Comment by AndyOGo** [comment]
hit the magic button^^

**Comment by davidknezic** [comment]
Awesome, will merge!

**Pull request merged**